compiler/rustc_borrowck/src/diagnostics/region_errors.rs

```diff
@@ -902,3 +945,40 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
         suggest_adding_lifetime_params(self.infcx.tcx, sub, ty_sup, ty_sub, diag);
     }
 }
+
+/// Suggests removing unnecessary lifetime annotations as `span`, assuming `span` is a path with
+/// lifetime annotations.
+///
+/// One subtle case is the use of `Self` which can impose unnecessary region constraints. Here we
+/// suggest replacing `Self` with `MyStruct<'_>`.
+pub(crate) fn suggest_relax_annotation<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    err: &mut Diagnostic,
+    body_did: DefId,
+    span: Span,
+) {
+    use rustc_lexer as lex;
+    fn tokenize(mut input: &str) -> impl Iterator<Item = (lex::TokenKind, &str)> {
```

Comment:
This makes me somewhat uncomfortable..

Reply:
Does it really matter? I mean it's only in the error path.

Reply:
Yeah, I think it does. There are lots of ways you can get spans to point at random things that aren't paths, like macro expansions, parser error recovery paths, etc. A lot of the error handling logic we have in borrowck is super fragile, and I think invoking the lexer of all things in borrowck is suspicious. How much effort would it be to pass down something like a hir id? or a rustc_middle Ty?

Reply:
I don't think passing Ty makes a difference, I'm trying to detect the use of `Self` here.
About the HirId, I'm more worried about the correctness of such approach, the mir/thir would reference hir and that's an abstraction leakage IMO.

Probably a better thing can be done here is to compute `uses_self: bool` field for each type annotation when building thir/mir. Bu this may be out of scope for this PR because I'm targeting a beta-backport. Is it ok to try this in a followup PR? or is there a chance that this gets backported anyway?